operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert(ctx),
+				ipam.svcOnDelete(ctx),
+			)
+		case err := <-errChan:
+			ipam.Logger.WithError(err).Error("Resource error")
+		}
+	}
+}
+
+func (ipam *LBIPAM) init(
+	ctx context.Context,
+	poolChan <-chan resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool],
+	svcChan <-chan resource.Event[*core_v1.Service],
+) {
+	// First sync all pools
+	poolSync := false
+	for !poolSync {
+		select {
+		case event := <-poolChan:
+			event.Handle(
+				func(s resource.Store[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]) error {
+					ipam.settleConflicts(ctx)
+					poolSync = true
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					ipam.handleNewPool(ctx, clbi)
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					ipam.Logger.Error("Got a Pool deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+
+	// Then sync all services
+	svcSync := false
+	for !svcSync {
+		select {
+		case event := <-svcChan:
+			event.Handle(
+				func(s resource.Store[*core_v1.Service]) error {
+					ipam.satisfyServices(ctx)
+					ipam.updateAllPoolCounts(ctx)
+					svcSync = true
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					ipam.handleUpsertService(ctx, svc)
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					ipam.Logger.Error("Got a Service deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+}
+
+func (ipam *LBIPAM) poolOnUpsert(ctx context.Context) func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	return func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+		// Deep copy so we get a version we are allowed to update
+		pool = pool.DeepCopy()
+
+		if _, exists := ipam.PoolStore.GetByUID(pool.GetUID()); exists {
+			ipam.Logger.Debugf("Updating pool '%s'", pool.GetName())
+			ipam.handlePoolModified(ctx, pool)
+		} else {
+			ipam.Logger.Debugf("Adding pool '%s'", pool.GetName())
+			ipam.handleNewPool(ctx, pool)
+		}
+
+		ipam.settleConflicts(ctx)
+		ipam.satisfyServices(ctx)
+		ipam.updateAllPoolCounts(ctx)
+
+		return nil
+	}
+}
+
+func (ipam *LBIPAM) poolOnDelete(ctx context.Context) func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	return func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+		ipam.handlePoolDeleted(ctx, pool)
+		ipam.settleConflicts(ctx)
+		ipam.satisfyServices(ctx)
+		ipam.updateAllPoolCounts(ctx)
+		return nil
+	}
+}
+
+func (ipam *LBIPAM) svcOnUpsert(ctx context.Context) func(k resource.Key, svc *core_v1.Service) error {
+	return func(k resource.Key, svc *core_v1.Service) error {
+		// Deep copy so we get a version we are allowed to update
+		svc = svc.DeepCopy()
```

Comment:
Hmm, should maybe make it such that `Resource[T]` always gives a unique copy. Would suck to debug a production issue because of mutation of a cache.Store object. Thoughts? @aanm would this be too costly default?

Reply:
>@aanm would this be too costly default?

@joamaki yes, we should likely understand why we need to do a deep copy in the first place. @dylandreimerink from a quick glance of the code it doesn't look like we are changing writing to any field in this resource so don't need to do a deep copy.